native/src/main/java/io/ballerina/stdlib/data/csvdata/csv/CsvTraversal.java

```diff
@@ -0,0 +1,802 @@
+                        }
+                        for (Type constituentType : ((IntersectionType) fieldType).getConstituentTypes()) {
+                            constituentType = TypeUtils.getReferredType(constituentType);
+                            if (constituentType.getTag() == TypeTags.READONLY_TAG) {
+                                continue;
+                            }
+                            return CsvCreator.constructReadOnlyValue(getFieldValue(constituentType,
+                                    csvMember, true));
+                        }
+                        break;
+                    default:
+                        throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_EXPECTED_TYPE, type);
+                }
+            }
+            return UnMappedValue.createUnMappedValue();
+        }
+
+        private void addRestField(Type type, BString key, Object csvMember) {
+            Object value = getFieldValue(type, csvMember, false);
+            if (!(value instanceof UnMappedValue)) {
+                ((BMap<BString, Object>) currentCsvNode).put(key, value);
+            }
+        }
+
+        private void addCurrentFieldValue(Type type, Object recValue, BString key, boolean isMapType) {
+            Object value = getFieldValue(type, recValue, false);
+            if (!(value instanceof UnMappedValue || value instanceof SkipMappedValue)) {
+                ((BMap<BString, Object>) currentCsvNode).put(StringUtils.fromString(fieldNames.pop()), value);
+                return;
+            }
+
+            if (isMapType || value instanceof SkipMappedValue) {
+                return;
+            }
+            throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_TYPE_FOR_FIELD, recValue, key);
+        }
+
+        public void addValuesToArrayType(Object arrayValue, Type type, int index,
+                                                 Object currentCsvNode, CsvConfig config) {
+            Object value = getFieldValue(type, arrayValue, false);
+            boolean isArrayType = type instanceof ArrayType;
+            if (!(value instanceof UnMappedValue)) {
+                if (isArrayType) {
+                    ArrayType arrayType = (ArrayType) TypeUtils.getType(type);
+                    if (arrayType.getState() == ArrayType.ArrayState.CLOSED &&
+                            arrayType.getSize() - 1 < index) {
+                        return;
+                    }
+                }
+                ((BArray) currentCsvNode).add(index, value);
+                return;
+            }
+            throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_TYPE_FOR_ARRAY, arrayValue, index, type);
+        }
+
+        private void setRootCsvNodeForNonUnionArrays(Type referredType, Type type) {
+            referredType = TypeUtils.getReferredType(referredType);
+            if (referredType.getTag() == TypeTags.ARRAY_TAG) {
+                ArrayType arrayType = (ArrayType) referredType;
```

Comment:
Given you are anyway going to cast to `ArrayType` in the dominant branch which is going to trigger a type check anyway it is more efficient to just to the type check (unless there are `ArrayType` that have a different tag which I don't think to be the case)
```suggestion
            if (referredType instanceOf ArrayType arrayType) {
```